### DISCUSSION TITLE
How to read large string from serial and split them

### DISCUSSION THREAD

**Discussion by mediaessenz** [arendst/Tasmota#20854] [Support request] [answered]
Hello,

I already discover this discussion, but did not found a solution for my case:
https://github.com/arendst/Tasmota/discussions/16497

I want to read data coming from a serial connection to my heating system.
It is 10 year old system from an austrian brand (Hargassner) and send some simple, but long text messages.

A message could be something like this:
pm 10 0.0 54 71 15.9 11.7 37.5 140.0 36.3 0.0 52 125 0 46 100.0 100.0 0 0.0 0.0 0.0 0 8.3 0 67 20 140 140 140.0 140.0 0.0 0.0 100.0 100.0 140 140 140.0 140.0 0.0 0.0 100.0 100.0 140 0.00 0.00 0.0 0.00 0.00 0.00 140.00 0 0 0 202 218 234 250 200 216 232 248 1 6 200 1 0 0 0 0 0 0 0 0 0 0 0 0 10 91 0 cb ffff 900 900 0 \n\rtm 2024-03-03 14:19:44

It's around 360 chars long and contains some special chars to separate the data from a timestamp.
The data part starts with the string "pm" and ends with a \n\r.

I'm interested in some of these values (all of them stuck in the first 128 bytes)  and would like to fire mqtt messages if they change.

Before I gave the tasmota scripting language a try, I just sent the whole string as a mqtt message to my broker and did the extraction of the interesting values in a node red instance using javascript. This works so far, but it's not really smart, since the heating system sends a data package every second.

After reading the scripting documentation and playing around with the possibilities I stuck now in some problems.
I could imagine that the restriction of the max length of a variable is a problem, but I'm not sure.

Here is the current state of my script:
```
>D 128
res=0
rstr=""
>B
res=so(3 -1 19200 8N1 1024)
>S
res=sa()
if res>0 {
 rstr=sb(sr(13) 0 128)
 if st(rstr ' ' 1)=='pm' {
  ; this should output the received data (until the 128 byte) starting with pm
  ; but it is changes all the time
  print received %rstr%
  ; this should e.g. output the 3rd value (54) a data string look like this
  ; pm 10 0.0 54 71 15.9 11.7 37.5 140.0 36.3 0.0 52 125 0 ....
  print %st(rstr ' ' 4)%
 }
}
```
This are my additional settings from the user_config_override.h
```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif

#ifdef USE_RULES
#undef USE_RULES
#endif

#ifndef USE_SCRIPT_WEB_DISPLAY
#define USE_SCRIPT_WEB_DISPLAY
#endif

#ifndef USE_SCRIPT_JSON_EXPORT
#define USE_SCRIPT_JSON_EXPORT
#endif

#ifndef USE_SCRIPT_SUB_COMMAND
#define USE_SCRIPT_SUB_COMMAND
#endif

#ifndef USE_SCRIPT_SERIAL
#define USE_SCRIPT_SERIAL
#endif

#define SCRIPT_MAXSSIZE 128
```

The hardware I'm currently using is an ESP-01.

**Comment by mediaessenz** [comment]
Meanwhile I found a solution:
```
>D 128
;kessel ist
tki=0
;kessel soll
tks=0
;vorlauf ist
tvi=0
;vorlauf soll
tvs=0
;aussen mittel
tam=0
;aussen gemessen
tag=0
;rauchgas
trg=0
;co2
co2=0
;boiler
tb=0
p1=""
data=""
msg="Keine"

>E
;Executed when a Tasmota MQTT RESULT message is received
;read value from serial
data=SerialReceived
p1=st(data ' ' 1)
print p1 %p1%
if (p1=="pm") {
 ;if data starts with „pm“ split values into array and assign them to the corresponding variables
 data=sb(data 3 100)
 print received parameter: %data%
 tki=st(data ' ' 3)
 print Kessel ist: %tki%
 tks=st(data ' ' 14)
 print Kessel soll: %tks%
 tvi=st(data ' ' 7)
 print Vorlauf ist: %tvi%
 tvs=st(data ' ' 9)
 print Vorlauf soll: %tvs%
 tam=st(data ' ' 5)
 print Aussentemperatur (mittel): %tam%
 tag=st(data ' ' 6)
 print Aussentemperatur (gem.): %tag%
 trg=st(data ' ' 4)
 print Rauchgas Temperatur: %trg%
 co2=st(data ' ' 2)
 print CO2: %co2%
 tb=st(data ' ' 11)
 print Boiler Temperatur: %tb%
 ;additional send mqtt SENSOR message (e.g %topic%/SENSOR {„Timestamp“:“%tstamp%“, „kesselIst“: 40.0}) if value has changed
 =>Publish tele/%topic%/SENSOR {"Time":"%tstamp%", "kesselIst": %tki%, "kesselSoll": %tks%, "vorlaufIst": %tvi%, "vorlaufSoll": %tvs%, "aussenMittel": %tam%, "aussenGemessen": %tag%, "rauchgas": %trg%, "co2": %co2%, "boiler": %tb%}
}

eres=1
```

The only problem I still have: additional to the published mqtt command inside my code the whole data string is published as well:
```
RESULT {"SerialReceived":"pm 0 18.3 36 36 6.0 7.4 34.5 140.0 0.0 0.0 54 125 0 0 100.0 100.0 0 0.0 0.0 0.0 0 8.8 0 67 30 140 140 140.0 140.0 0.0 0.0 100.0 100.0 140 140 140.0 140.0 0.0 0.0 100.0 100.0 140 0.00 0.00 0.0 0.00 0.00 0.00 140.00 0 0 0 202 218 234 250 200 216 232 248 1 6 200 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 ffff 900 900 0 \n\rtm 2024-03-03 23:31:13"}
```
The added command `eres=1` which should prevent this has no effect.

Or do I misunderstand this command described here: https://tasmota.github.io/docs/Scripting-Language/#script-sections

How can I prevent this?

**Reply by sfromis** [reply]
Not related to scripting, the general approach is that serial bridge messages always are published, with the option for you to pick up the content, and process it. Thus there is no problem.

However, a few months ago, a feature was added, `SetOption147` allowing you to selectively disable that payload from being published.

**Comment by mediaessenz** [comment]
Thank you for answering!

Unfortunately SetOption147 only affects **S**SerialReceived and has no effect in my case.

Do you know a way to do the same with SerialReceived?

Since the number of ports of my ESP-01 is very limited (just four), I really would like to leave the RX wire where it is.

Do you know whether it's possible to change "RX GPIO3" in module configuration from "Serial Rx" to "SerBr Rx"?
Or is this GPIO3 always blocked for HW serial?

Another problem/question:
I just added a not yet wired "SerBr Rx" to GPIO2 inside the module configuration.
After restarting the device I went to the console and tried to configure the baudrate and protocol of this software serial, but it doesn't seems to work.
If I e.g. enter the command `SBaudrate 19200` I get a message like this: 
```
11:11:36.601 MQT: stat/hargassner_0700F6/RESULT = {"Command":"Unknown"}
```
To be sure this command is included in my build, I added this to my user_config_override.h:
```
#ifndef USE_SERIAL_BRIDGE
#define USE_SERIAL_BRIDGE
#endif
```
Do I miss something or is software serial just not supported by my device (ESP-01)?
If so, why is it listed in the module configuration?

**Reply by sfromis** [reply]
Serial Bridge is not operational without having both TX and RX pins. And while ESP-01 is basically the worst device to build on, it is the same ESP8266 and same features within the paucity of available pins.

If you want the worse performance of software serial (not recommended with long messages as you have), you could just try if it works on the pins for the better hardware serial. But using hardware serial is strongly recommended, and is basically the only use case ESP-01 is designed for.

**Reply by mediaessenz** [reply]
It looks, using the software serial is not really a option 😢 

Why there is no "SetOption147" for SerialReceived?
Is this just a missing feature or is it technical not possible?
I could imagine tasmota generates the mqtt message with the serial received data. Why should is be not possible to prevent this?

**Reply by sfromis** [reply]
In most cases the answer to "why" is "just because", meaning that lots of features are created to handle some type of situation, instead of trying to do a lot more. The MQTT messages are key to you being able to pickup the payloads and publish, and it is only in special narrow cases that the publish step can be skipped.

**Reply by mediaessenz** [reply, marked as answer]
All right.

Meanwhile I gave the software serial a try and it looks good for me. It's maybe not that stable, but for my use case it's okay.

I created a bridge between GPIO3 and GPIO0 and configured GPIO0 as SerBr RX and GPIO2 as SerBr TX.
Now (thanks to your hint!) I was able to configure the software serial to the same baudrate as the hardware serial (command: `SBaudrate 19200`).

After modifying my script a bit (I just changed `data=SerialReceived` to `data=SSerialReceived`), disabled hardware serial logging (command: `SerialLog 0`) and disable population of software serial received data (command: `SetOption147 1`) I got the desired result.
Sometimes I see the message `Serial buffer overrun` in the console, but that's fine.
The received string is around 360 bytes long, but the interested values stuck all in the first 100 bytes.

😄

**Reply by Jason2866** [reply]
Just was to write give Software Serial a try ;-) Worst case you miss now and than an info. Probably not a big deal for a heating sending the data every second.

**Reply by sfromis** [reply]
You could consider a custom build to expand:
```
#define SERIAL_BRIDGE_BUFFER_SIZE 256          // Serial Bridge receive buffer size (Default ESP8266 = 256, ESP32 = 800)
```

**Reply by mediaessenz** [reply]
Doesn't seems to make any different.
Set it to `#define SERIAL_BRIDGE_BUFFER_SIZE 384` and still get the buffer overflows.

But as wrote, I don't mind. 😉

**Reply by sfromis** [reply]
I forgot that there's also a `SerialBuffer` command for the setting, which means that the *default* you can set at build time only takes effect with a full reset. After that, the current value in settings persists. Try the command.
